File: src/chapters/linear-systems/content/code/forward-substitution.m
function y = forward_substitution(L, b)
% FORWARD_SUBSTITUTION  Solve a lower-triangular system L y = b.
    n = numel(b); y = zeros(n, 1);
    for i = 1:n
        y(i) = (b(i) - L(i, 1:i-1) * y(1:i-1)) / L(i, i);
    end
end

% --- Demo ---
L = [2 0 0; 1 3 0; -1 1 2]; b = [4; 5; -1];
disp(forward_substitution(L, b)');    % 2 1 0

File: src/chapters/linear-systems/content/code/forward-substitution.go
package main

import "fmt"

// Solve a lower-triangular system L y = b.
func forwardSubstitution(L [][]float64, b []float64) []float64 {
	n := len(b)
	y := make([]float64, n)
	for i := 0; i < n; i++ {
		s := b[i]
		for j := 0; j < i; j++ {
			s -= L[i][j] * y[j]
		}
		y[i] = s / L[i][i]
	}
	return y
}

func main() {
	L := [][]float64{{2, 0, 0}, {1, 3, 0}, {-1, 1, 2}}
	b := []float64{4, 5, -1}
	fmt.Println(forwardSubstitution(L, b)) // [2 1 0]
}
